h2o-genmodel-extensions/xgboost/src/main/java/hex/genmodel/algos/xgboost/XGBoostMojoModel.java

```diff
@@ -42,9 +46,56 @@ public XGBoostMojoModel(String[] columns, String[][] domains, Booster _booster,
             _booster, _nums, _cats, _catOffsets, _useAllFactorLevels,
             _nclasses, _priorClassDistrib, _defaultThreshold, _sparse);
   }
-  public static final double[] score0(double[] doubles, double offset, double[] preds,
-      Booster _booster, int _nums, int _cats, int[] _catOffsets, boolean _useAllFactorLevels,
-                                      int nclasses, double[] _priorClassDistrib, double _defaultThreshold, boolean _sparse) {
+
+  public static double[][] bulkScore0(double[][] doubles, double[] offsets, double[][] preds,
+                                      Booster _booster, int _nums, int _cats,
+                                      int[] _catOffsets, boolean _useAllFactorLevels,
+                                      int nclasses, double[] _priorClassDistrib,
+                                      double _defaultThreshold, boolean _sparse) {
+    if (offsets != null) throw new UnsupportedOperationException("Unsupported: offset != null or only 0s");
+    float[][] floats;
+    int cats = _catOffsets == null ? 0 : _catOffsets[_cats];
+    // convert dense doubles to expanded floats
+    floats = new float[doubles.length][_nums + cats]; //TODO: use thread-local storage
+    for(int i = 0; i < doubles.length; i++) {
+      GenModel.setInput(doubles[i], floats[i], _nums, _cats, _catOffsets, null, null, _useAllFactorLevels, _sparse /*replace NA with 0*/);
+    }
+    float[][] out = null;
+    try {
+      Map<String, String> rabitEnv = new HashMap<>();
+      rabitEnv.put("DMLC_TASK_ID", "0");
+      Rabit.init(rabitEnv);
+      DMatrix dmat = new DMatrix(floats,doubles.length,floats[0].length, _sparse ? 0 : Float.NaN);
+//      dmat.setWeight(new float[]{(float)weight});
+      out = _booster.predict(dmat);
+      Rabit.shutdown();
+    } catch (XGBoostError xgBoostError) {
+      throw new IllegalStateException("Failed XGBoost prediction.", xgBoostError);
+    }
+
+    for(int r = 0; r < out.length; r++) {
+      if (nclasses > 2) {
+        for (int i = 0; i < out[0].length; ++i)
+          preds[r][1 + i] = out[r][i];
+//      if (_balanceClasses)
```

Comment:
Maybe remove the comments containing code if they are out-dated, not important ?